src/sql/src/query_model/mod.rs

```diff
@@ -0,0 +1,417 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::cell::RefCell;
+use std::collections::BTreeSet;
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+mod dot_generator;
+mod hir_generator;
+mod scalar_expr;
+#[cfg(test)]
+mod test;
+
+pub use scalar_expr::*;
+
+pub type QuantifierId = usize;
+pub type BoxId = usize;
+pub type QuantifierSet = BTreeSet<QuantifierId>;
+
+/// A Query Graph Model instance represents a SQL query.
+#[derive(Debug)]
+pub struct Model {
+    /// The ID of the box representing the entry-point of the query.
+    pub top_box: BoxId,
+    /// All boxes in the query graph model.
+    boxes: HashMap<BoxId, Box<RefCell<QueryBox>>>,
+    /// Used for asigning unique IDs to query boxes.
+    next_box_id: usize,
```

Comment:
There already exists an `IdGen` struct, albeit in `MirRelationExpr`. https://github.com/MaterializeInc/materialize/blob/c9080b3065442b5d04dfa2fca9ad07f0067f7ea8/src/expr/src/relation/mod.rs#L1485
Perhaps `IdGen` should be moved to `src/ore` so it can be repeatedly used.

Reply:
I moved and used IdGen here.